### DISCUSSION TITLE
Sharing Channels Between Threads Using Locks

### DISCUSSION THREAD

**Discussion by jorgasto** [ruby-amqp/bunny#667] [Q&A] [answered]
Hello!

In the Bunny [documentation](https://github.com/ruby-amqp/bunny/blob/main/docs/guides/concurrency.md) I cannot quite tell if you guys are saying channels should only be used by one thread EVER or one thread at a time (i.e. using synchronization).

e.g. is it possible for the following script to cause the concurrency issues mentioned in the docs?

``` ruby
class ChannelPool
  def initialize(channel)
     @channel = channel
     @mutex = Mutex.new
  end
  
  def take
    @mutex.synchronize do
      yield @channel
    end
  end
end

pool = ChannelPool.new

Thread.new do 
   while true do
      pool.take do |channel|
        channel.exchange('foo', type: :fanout, durable: true, auto_delete: false)
     end
   end
end

Thread.new do 
   while true do
      pool.take do |channel|
        channel.exchange('bar', type: :fanout, durable: true, auto_delete: false)
     end
   end
end

```

**Comment by michaelklishin** [comment, marked as answer]
This code assumes that `Bunny::Channel#exchange` calls cannot result in any kind of interleaving of frames on the wire. Right now this is generally true, and `exchange.declare` is a single protocol frame, and Bunny writes frames to the socket with synchronisation.

However, what's the point of declaring exchanges if you are not going to publish anything? And `Bunny::Channel#publish` produces two or more frames, depending on message size:

```
[basic.publish] [content header] [body]*
```

So even with synchronised writes of individual frames, you have to be really careful about not ending up with a sequence like this

```
[basic.publish 1] [basic.publish 2] [content header 2] [body 2]* [content header 1] [body 1]*
```

This will cause an unrecoverable connection exception.

It is much safer to simply not share channel instances between threads.

**Reply by jorgasto** [reply]
Thanks for clearing that up @michaelklishin!

**Reply by sanghyukkang** [reply]
Thank you michaelklishin, but I would like to know how below two cases could be different regarding the race condition on the frame sequence.

Case 1. Single thread publishes two messages in sequence.
```ruby
channel = conn.create_channel
channel.publish(msg1)
channel.publish(msg2)
```

Case 2. Two threads, each publishes one message in parallel but synchronized by a mutex, assuming that the two threads actually run concurrently. The point is that only one thread uses the channel at the same time.
```ruby
mutex = Mutex.new
channel = conn.create_channel

Thread.new do # thread 1
  mutex.lock
  channel.publish(msg1)
  mutex.unlock
end

Thread.new do # thread 2
  mutex.lock
  channel.publish(msg2)
  mutex.unlock
end
```

**Reply by michaelklishin** [reply]
Your code does not handle any possible exceptions that `Bunny::Channel#publish` may throw. Consider using a block version (`Mutex#synchronized`).

You can set Bunny log level to debug and see what frames it sends.

**Reply by sanghyukkang** [reply]
Thank you michael. The code was like a pseudo code just to compare single thread publishing and multi thread publishing. I was wondered how two message frames can be interlaced when the two `channel.publish` method calls are completely synchronized. There is no concurrent execution of `channel.publish` even in Case 2. Second `channel.publish`, no matter which thread is the second, starts after first `channel.publish` has returned.

**Reply by michaelklishin** [reply]
There is concurrent execution in your own code (with synchronization, and yet), and thread scheduling order is non-deterministic.

Remove the synchronization and see for yourself. RabbitMQ will log a framing error soon enough

**Reply by sanghyukkang** [reply]
I know that there is concurrent execution if there is no synchronization. What I meant was that there wouldn't be actual concurrent execution of `channel.publish` method in Case 2 because of the synchronization. I'm only interested in the case where there is synchronization, whether or not there could be a race condition like the interlaced message frames even with the synchronization.

**Reply by michaelklishin** [reply]
Bunny does not introduce any concurrency from the moment `Bunny::Channel#publish` is invoked all the way up to socket write. So in the example above, other than the lack of exception handling, I don't see any obvious concurrency hazard scenarios.
